EIPS/eip-4361.md

```diff
@@ -0,0 +1,304 @@
+---
+eip: 4361
+title: Sign-In with Ethereum.
+description: Standardizes off-chain authentication for Ethereum accounts to establish sessions.
+author: Wayne Chang (@wyc), Brantly Millegan (@brantlymillegan), Nick Johnson (@Arachnid)
+resolution: TBD
+discussions-to: https://ethereum-magicians.org/t/eip-4361-sign-in-with-ethereum/7263
+status: Draft
+type: Standards Track
+category: ERC
+created: 2021-10-11
+requires: 55, 191, 137, 155, 1271, 1328
+---
+
+## Abstract
+Sign-In with Ethereum describes how Ethereum accounts authenticate with off-chain services by signing a standard message format parameterized by scope, session details, and security mechanisms (e.g., a nonce). The goals of this specification are to provide a self-custodied alternative to centralized identity providers, improve interoperability across off-chain services for Ethereum-based authentication, and provide wallet vendors a consistent machine-readable message format to achieve improved user experiences and consent management.
+
+## Motivation
+When signing in to popular non-blockchain services today, users will typically use identity providers (IdPs) that are centralized entities with ultimate control over users' identifiers, for example, large internet companies and email providers. Incentives are often misaligned between these parties. Sign-In with Ethereum offers a new self-custodial option for users who wish to assume more control and responsibility over their own digital identity.
+
+Already, many services support workflows to authenticate Ethereum accounts using message signing, such as to establish a cookie-based web session which can manage privileged metadata about the authenticating address. This is an opportunity to standardize the sign-in workflow and improve interoperability across existing services, while also providing wallet vendors a reliable method to identify signing requests as Sign-In with Ethereum requests for improved UX.
+
+This work is sponsored by the Ethereum Foundation and Ethereum Name Service (ENS). It is being developed in the open through a series of recorded community calls and public repositories, and its development is informed by over twenty user interviews with a focus on currently-in-production uses, related prior EIPs, and fits within product roadmaps.
+
+## Specification
+Sign-In with Ethereum works as follows:
+
+1. The wallet presents the user with a structured plaintext message or equivalent interface for signing with the [EIP-191](https://eips.ethereum.org/EIPS/eip-191) signature scheme (string prefixed with `\x19Ethereum Signed Message:\n<length of message>`). The `message` MUST incorporate an Ethereum `address`, `version` of the message, `uri` for scoping, `nonce` acceptable to the server, and `issued-at` timestamp.
```

Comment:
Does reusing `\x19Ethereum Signed Message:\n<length of message>` pose a security risk?

If an attacker on a visually similar phishing domain uses the existing `eth_signMessage` endpoint to sign a message, do we expect existing wallets to display a warning?

Reply:
I don't quite understand the question here related to the prefix (as mentioned in EIP-191) and domains?

If we want the wallet to display a warning then we'd need to require the wallet to parse a domain from the message and check this against the domain of the website in use. Should wallets do that? Sure, it would be nice and good UX as far as I'm concerned.

But how does this relate to the prefix?

Reply:
If you use a separate prefix, existing RPC endpoints like `eth_signMessage` will generally refuse to sign, so the wallets don't need to implement any special checks on _all_ signed messages, only on ones that use the new prefix.

If you use the same prefix, wallets have to scan all `eth_signMessage` bodies for messages that parse as EIP-4361 compatible, and then check the domain.

Reply:
@SamWilsn Thanks for your comments! We should encourage wallet vendors to display a separate interface once they have identified a valid SIWE message, but it's true they'd have to interpret the whole thing in the current form. Two possible improvements here, not mutually exclusive:

1. We could alternatively require the message to begin with `"EIP-4361. "` to allow quick prefix determination for parsers prior to attempting to parse any of the grammar, but I think the message should be short enough as not to pose a big risk. This would potentially set a precedence on how future signing requests will differentiate themselves, inadvertently setting a precedent for future EIP-191 extensions. It could also slightly degrade the user experience, but arguably also improve it by giving users a term that is likely to turn up with useful results in a search engine.

2. We can add a SHOULD clause to the spec that cannot be easily tested like: "Wallet vendors SHOULD display a confirmatory indicator if a valid EIP-4361 message is detected, and wallets MAY take reasonable steps to proactively identify similar but non-conformant messages to alert the user accordingly."

Thoughts on these?

Reply:
Also, note on the separate prefix--I don't think most wallets would be able to display the message cohesively if we removed the signing prefix `\x19Ethereum Signed Message:\n<length of message>` for wallets that don't make changes explicitly to support this spec. The goal is that applications should be able to create conforming implementations without forcing wallet vendors to do anything without substantially degrading the user experience. This is an important consideration for adoption, and at the same time your observation is valid.

Reply:
I'm not a fan of (1) since it adds extra gibberish to the message that will make users even more likely to just blindly sign.

Is the `wants you to sign in with your Ethereum account` portion fixed and required? If so, that's probably sufficient for wallets to scan for.

I do think (2) is a good security recommendation. Maybe something as simple as: "Wallet software SHOULD display a high friction confirmatory indicator if the string `wants you to sign in with your Ethereum account` appears anywhere in an EIP-191 message signing request unless the message is a valid according to EIP-4361."

Reply:
Thanks! I will make these changes, okay to add you as a contributor for your useful comments throughout?

Reply:
Sure, that's fine!

Reply:
I also found the `wants you to sign in with your Ethereum account` to be an "interesting" part of the specification.

How does this work with multiple languages in regards to the specification? Ideally the specification has no dependency on the English language and can be used in a universal setting. Can this be replaced with translated equivalents, and if so will wallet vendors be expected to parse for those strings as well? 

This is kind of secondary, but I found it just a little awkward to parse the current message template. Here is just a quick, naive approach. 

https://gist.github.com/kamescg/b0f2875881ff15ec976f6976cedca885

Reply:
> Maybe something as simple as: "Wallet software SHOULD display a high friction confirmatory indicator if the string wants you to sign in with your Ethereum account appears anywhere in an EIP-191 message signing request unless the message is a valid according to EIP-4361."

The best solution definitely feels like parsing the `eth_signMessage` to detect message `types` is the most flexible long-term solution. Also, it's hard to believe this will be the last EIP to use string parsing to detect the message `intention`.

Because of this I feel like detecting such a "generic message" like `wants you to sign in with your Ethereum account` is not the most scalable approach to dynamically detecting all incoming `eth_signMessage` potential future types.

In my opinion it would probably be best to drop the `wants you to sign in with your Ethereum account` because that's implied in the singing 😉  and replace it with a `type` and `language` keys that can be parsed by incoming wallets and also be used by future standards who need to have a `StructuredDataIsh` interface while still using the `eth_signMessage` call.

Easy to parse and identify.

And the intention message maybe has to be replaced with `Domain` and `Address` keys?

Also just curious as to why the `statement` param isn't a key/value pair like everything else? Is a bit of an odd data structure when the message has to be dynamically parsed and found by searching for the two surrounding `\n` on empty lines.

**Proposed Changes**
```
type: EIP-4361
language: en
Domain: http://localhost:3000 
Address: 0x000000000000000000000000000000000000dEaD
Statement: http://opensea.io/tos
URI: http://localhost:3000/authenticate
Version: 1
Chain ID: 1
Nonce: 12345678
Issued At: 2021-11-20T10:10:10Z
Not Before: 2021-11-19T10:10:10Z
Expiration Time: 2021-11-21T10:10:10Z
Request ID: 2
Resources:
 https://ipfs.com/hash
```

Reply:
Hi Kames, the example you present is indeed more normalized, but I am concerned about readability to end users. We gathered feedback from over 30+ teams in the space over the past several weeks, and they were okay with the proposed format. We welcome your independent research to see if they prefer the format you propose for their users over the current one if you wish to move it forward.

Additionally, I would consider solving for a "scalable approach to dynamically detecting all incoming eth_signMessage potential future types" is out of scope for this spec.

Thanks again for your comments, and I look forward to continuing the conversation in ETH Magicians if you continue to feel strongly about this.